File: android/src/main/kotlin/com/changteng/agora_rtc/AgoraRtcEngineMediaActions.kt
package com.changteng.agora_rtc

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

class AgoraRtcEngineMediaActions(
  private val engineHolder: AgoraRtcEngineHolder
) {
  fun handleMuteAllRemoteAudioStreams(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val muted = args["muted"] as? Boolean
    if (muted == null) {
      result.error("INVALID_ARGS", "muted 为空", null)
      return
    }
    val code = engine.muteAllRemoteAudioStreams(muted)
    result.success(code)
  }

  fun handleMuteAllRemoteVideoStreams(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val muted = args["muted"] as? Boolean
    if (muted == null) {
      result.error("INVALID_ARGS", "muted 为空", null)
      return
    }
    val code = engine.muteAllRemoteVideoStreams(muted)
    result.success(code)
  }

  fun handleMuteRemoteAudioStream(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val uid = (args["uid"] as? Number)?.toInt()
    val muted = args["muted"] as? Boolean
    if (uid == null || muted == null) {
      result.error("INVALID_ARGS", "uid 或 muted 为空", null)
      return
    }
    val code = engine.muteRemoteAudioStream(uid, muted)
    result.success(code)
  }

  fun handleMuteRemoteVideoStream(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val uid = (args["uid"] as? Number)?.toInt()
    val muted = args["muted"] as? Boolean
    if (uid == null || muted == null) {
      result.error("INVALID_ARGS", "uid 或 muted 为空", null)
      return
    }
    val code = engine.muteRemoteVideoStream(uid, muted)
    result.success(code)
  }

  fun handleMuteLocalAudioStream(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val muted = args["muted"] as? Boolean
    if (muted == null) {
      result.error("INVALID_ARGS", "muted 为空", null)
      return
    }
    val code = engine.muteLocalAudioStream(muted)
    result.success(code)
  }

  fun handleMuteLocalVideoStream(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val muted = args["muted"] as? Boolean
    if (muted == null) {
      result.error("INVALID_ARGS", "muted 为空", null)
      return
    }
    val code = engine.muteLocalVideoStream(muted)
    result.success(code)
  }

  fun handleSetRemoteVideoStreamType(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val uid = (args["uid"] as? Number)?.toInt()
    val streamType = (args["streamType"] as? Number)?.toInt()
    if (uid == null || streamType == null) {
      result.error("INVALID_ARGS", "uid 或 streamType 为空", null)
      return
    }
    val code = engine.setRemoteVideoStreamType(uid, streamType)
    result.success(code)
  }

  fun handleEnableVideo(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val enabled = args["enabled"] as? Boolean
    if (enabled == null) {
      result.error("INVALID_ARGS", "enabled 为空", null)
      return
    }
    val code = if (enabled) engine.enableVideo() else engine.disableVideo()
    result.success(code)
  }

  fun handleEnableLocalVideo(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val enabled = args["enabled"] as? Boolean
    if (enabled == null) {
      result.error("INVALID_ARGS", "enabled 为空", null)
      return
    }
    val code = engine.enableLocalVideo(enabled)
    result.success(code)
  }

  fun handleStartPreview(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val code = engine.startPreview()
    result.success(code)
  }

  fun handleStopPreview(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val code = engine.stopPreview()
    result.success(code)
  }

  fun handleTakeSnapshot(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val uid = (args["uid"] as? Number)?.toInt()
    val filePath = args["filePath"] as? String
    if (uid == null || filePath.isNullOrBlank()) {
      result.error("INVALID_ARGS", "uid 或 filePath 为空", null)
      return
    }
    val code = engine.takeSnapshot(uid, filePath)
    result.success(code)
  }
}

File: ios/Classes/AgoraRtcEngineMediaActions.swift
import Flutter
import AgoraRtcKit

final class AgoraRtcEngineMediaActions {
  private let engineHolder: AgoraRtcEngineViewHolder

  init(engineHolder: AgoraRtcEngineViewHolder) {
    self.engineHolder = engineHolder
  }

  func handleMuteAllRemoteAudioStreams(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let muted = args["muted"] as? Bool else {
      result(FlutterError(code: "INVALID_ARGS", message: "muted 为空", details: nil))
      return
    }
    let code = engine.muteAllRemoteAudioStreams(muted)
    result(code)
  }

  func handleMuteAllRemoteVideoStreams(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let muted = args["muted"] as? Bool else {
      result(FlutterError(code: "INVALID_ARGS", message: "muted 为空", details: nil))
      return
    }
    let code = engine.muteAllRemoteVideoStreams(muted)
    result(code)
  }

  func handleMuteRemoteAudioStream(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let uid = (args["uid"] as? NSNumber)?.uintValue else {
      result(FlutterError(code: "INVALID_ARGS", message: "uid 为空", details: nil))
      return
    }
    guard let muted = args["muted"] as? Bool else {
      result(FlutterError(code: "INVALID_ARGS", message: "muted 为空", details: nil))
      return
    }
    let code = engine.muteRemoteAudioStream(uid, mute: muted)
    result(code)
  }

  func handleMuteRemoteVideoStream(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let uid = (args["uid"] as? NSNumber)?.uintValue else {
      result(FlutterError(code: "INVALID_ARGS", message: "uid 为空", details: nil))
      return
    }
    guard let muted = args["muted"] as? Bool else {
      result(FlutterError(code: "INVALID_ARGS", message: "muted 为空", details: nil))
      return
    }
    let code = engine.muteRemoteVideoStream(uid, mute: muted)
    result(code)
  }

  func handleMuteLocalAudioStream(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let muted = args["muted"] as? Bool else {
      result(FlutterError(code: "INVALID_ARGS", message: "muted 为空", details: nil))
      return
    }
    let code = engine.muteLocalAudioStream(muted)
    result(code)
  }

  func handleMuteLocalVideoStream(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let muted = args["muted"] as? Bool else {
      result(FlutterError(code: "INVALID_ARGS", message: "muted 为空", details: nil))
      return
    }
    let code = engine.muteLocalVideoStream(muted)
    result(code)
  }

  func handleSetRemoteVideoStreamType(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let uid = (args["uid"] as? NSNumber)?.uintValue else {
      result(FlutterError(code: "INVALID_ARGS", message: "uid 为空", details: nil))
      return
    }
    guard let streamType = (args["streamType"] as? NSNumber)?.intValue else {
      result(FlutterError(code: "INVALID_ARGS", message: "streamType 为空", details: nil))
      return
    }
    let code = engine.setRemoteVideoStream(uid, type: AgoraVideoStreamType(rawValue: streamType) ?? .high)
    result(code)
  }

  func handleEnableVideo(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let enabled = args["enabled"] as? Bool else {
      result(FlutterError(code: "INVALID_ARGS", message: "enabled 为空", details: nil))
      return
    }
    let code = enabled ? engine.enableVideo() : engine.disableVideo()
    result(code)
  }

  func handleEnableLocalVideo(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let enabled = args["enabled"] as? Bool else {
      result(FlutterError(code: "INVALID_ARGS", message: "enabled 为空", details: nil))
      return
    }
    let code = engine.enableLocalVideo(enabled)
    result(code)
  }

  func handleStartPreview(result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    let code = engine.startPreview()
    result(code)
  }

  func handleStopPreview(result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    let code = engine.stopPreview()
    result(code)
  }

  func handleTakeSnapshot(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let uid = (args["uid"] as? NSNumber)?.uintValue else {
      result(FlutterError(code: "INVALID_ARGS", message: "uid 为空", details: nil))
      return
    }
    guard let filePath = args["filePath"] as? String, !filePath.isEmpty else {
      result(FlutterError(code: "INVALID_ARGS", message: "filePath 为空", details: nil))
      return
    }
    let code = engine.takeSnapshot(Int(uid), filePath: filePath)
    result(code)
  }
}
